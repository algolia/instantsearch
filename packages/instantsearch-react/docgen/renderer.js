import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
