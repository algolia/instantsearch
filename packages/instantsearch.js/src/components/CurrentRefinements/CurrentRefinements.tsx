/** @jsx h */

import { cx } from '@algolia/ui-components-shared';
import { h } from 'preact';

import { isSpecialClick, capitalize } from '../../lib/utils';

import type {
  CurrentRefinementsConnectorParamsItem,
  CurrentRefinementsConnectorParamsRefinement,
} from '../../connectors/current-refinements/connectCurrentRefinements';
import type { ComponentCSSClasses } from '../../types';
import type { CurrentRefinementsCSSClasses } from '../../widgets/current-refinements/current-refinements';

export type CurrentRefinementsComponentCSSClasses =
  ComponentCSSClasses<CurrentRefinementsCSSClasses>;

export type CurrentRefinementsProps = {
  items: CurrentRefinementsConnectorParamsItem[];
  cssClasses: CurrentRefinementsComponentCSSClasses;
  canRefine: boolean;
};

const createItemKey = ({
  attribute,
  value,
  type,
  operator,
}: CurrentRefinementsConnectorParamsRefinement): string =>
  [attribute, type, value, operator]
    .map((key) => key)
    .filter(Boolean)
    .join(':');

const handleClick = (callback: () => void) => (event: any) => {
  if (isSpecialClick(event)) {
    return;
  }

  event.preventDefault();
  callback();
};

const CurrentRefinements = ({
  items,
  cssClasses,
  canRefine,
}: CurrentRefinementsProps) => (
  <div
    className={cx(cssClasses.root, !canRefine && cssClasses.noRefinementRoot)}
  >
    <ul className={cssClasses.list}>
      {items.map((item, index) => (
        <li
          key={`${item.indexName}-${item.attribute}-${index}`}
          className={cssClasses.item}
        >
          <span className={cssClasses.label}>{capitalize(item.label)}:</span>

          {item.refinements.map((refinement) => (
            <span
              key={createItemKey(refinement)}
              className={cssClasses.category}
            >
              <span className={cssClasses.categoryLabel}>
                {refinement.attribute === 'query' ? (
                  <q>{refinement.label}</q>
                ) : (
                  refinement.label
                )}
              </span>

              <button
                className={cssClasses.delete}
                onClick={handleClick(item.refine.bind(null, refinement))}
              >
                âœ•
              </button>
            </span>
          ))}
        </li>
      ))}
    </ul>
  </div>
);

export default CurrentRefinements;
