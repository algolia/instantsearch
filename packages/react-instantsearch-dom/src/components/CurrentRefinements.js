import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { translatable } from 'react-instantsearch-core';

import { createClassNames } from '../core/utils';

const cx = createClassNames('CurrentRefinements');

export const CurrentRefinements = ({
  items,
  canRefine,
  refine,
  translate,
  className,
}) => (
  <div className={classNames(cx('', !canRefine && '-noRefinement'), className)}>
    <ul className={cx('list', !canRefine && 'list--noRefinement')}>
      {items.map((item) => (
        <li key={item.label} className={cx('item')}>
          <span className={cx('label')}>{item.label}</span>
          {item.items ? (
            item.items.map((nest) => (
              <span key={nest.label} className={cx('category')}>
                <span className={cx('categoryLabel')}>{nest.label}</span>
                <button
                  className={cx('delete')}
                  onClick={() => refine(nest.value)}
                >
                  {translate('clearFilter', nest)}
                </button>
              </span>
            ))
          ) : (
            <span className={cx('category')}>
              <button
                className={cx('delete')}
                onClick={() => refine(item.value)}
              >
                {translate('clearFilter', item)}
              </button>
            </span>
          )}
        </li>
      ))}
    </ul>
  </div>
);

const itemPropTypes = PropTypes.arrayOf(
  PropTypes.shape({
    label: PropTypes.string.isRequired,
    value: PropTypes.func.isRequired,
    items: (...args) => itemPropTypes(...args),
  })
);

CurrentRefinements.propTypes = {
  items: itemPropTypes.isRequired,
  canRefine: PropTypes.bool.isRequired,
  refine: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  className: PropTypes.string,
};

CurrentRefinements.defaultProps = {
  className: '',
};

export default translatable({
  clearFilter: 'âœ•',
})(CurrentRefinements);
