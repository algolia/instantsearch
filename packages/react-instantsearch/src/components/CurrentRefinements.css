.root {
}

.items {
}

.item {
}

.itemLabel {
}

.itemParent {
}

.itemClear {
}