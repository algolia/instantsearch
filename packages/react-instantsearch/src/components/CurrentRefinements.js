import React, {PropTypes, Component} from 'react';

import translatable from '../core/translatable';
import classNames from './classNames.js';

const cx = classNames('CurrentRefinements');

class CurrentRefinements extends Component {
  static propTypes = {
    translate: PropTypes.func.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
    })).isRequired,
    refine: PropTypes.func.isRequired,
    transformItems: PropTypes.func,
  };

  render() {
    const {translate, items, refine} = this.props;

    if (items.length === 0) {
      return null;
    }

    return (
      <div {...cx('root')}>
        <div {...cx('items')}>
          {items.map(item =>
            <div
              key={item.label}
              {...cx('item', item.items && 'itemParent')}
            >
              <span {...cx('itemLabel')}>
                {item.label}
              </span>
              {item.items ?
                item.items.map(nestedItem =>
                  <div
                    key={nestedItem.label}
                    {...cx('item')}
                  >
                  <span {...cx('itemLabel')}>
                  {nestedItem.label}
                  </span>
                    <button
                      {...cx('itemClear')}
                      onClick={refine.bind(null, nestedItem.value)}
                    >
                      {translate('clearFilter', nestedItem)}
                    </button>
                  </div>) :
                <button
                  {...cx('itemClear')}
                  onClick={refine.bind(null, item.value)}
                >
                  {translate('clearFilter', item)}
                </button>}
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default translatable({
  clearFilter: 'âœ•',
})(CurrentRefinements);
