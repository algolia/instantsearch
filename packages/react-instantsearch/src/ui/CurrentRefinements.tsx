import React from 'react';

import { capitalize, cx, isModifierClick } from './lib';

import type { CurrentRefinementsConnectorParamsItem } from 'instantsearch.js/es/connectors/current-refinements/connectCurrentRefinements';

export type CurrentRefinementsProps = React.ComponentProps<'div'> & {
  classNames?: Partial<CurrentRefinementsClassNames>;
  items?: Array<
    Pick<
      CurrentRefinementsConnectorParamsItem,
      'label' | 'refinements' | 'refine'
    > &
      Record<string, unknown>
  >;
  hasRefinements?: boolean;
};

export type CurrentRefinementsClassNames = {
  /**
   * Class names to apply to the root element
   */
  root: string;
  /**
   * Class names to apply to the root element when there are no refinements possible
   */
  noRefinementRoot: string;
  /**
   * Class names to apply to the list element
   */
  list: string;
  /**
   * Class names to apply to the list element when there are no refinements possible
   */
  noRefinementList: string;
  /**
   * Class names to apply to each refinement
   */
  item: string;
  /**
   * Class names to apply to the label of each refinement
   */
  label: string;
  /**
   * Class names to apply to the container of each refinement's value
   */
  category: string;
  /**
   * Class names to apply to the text element of each refinement's value
   */
  categoryLabel: string;
  /**
   * Class names to apply to the each refinement's delete button
   */
  delete: string;
};

export function CurrentRefinements({
  classNames = {},
  items = [],
  hasRefinements = false,
  ...props
}: CurrentRefinementsProps) {
  return (
    <div
      {...props}
      className={cx(
        'ais-CurrentRefinements',
        classNames.root,
        !hasRefinements &&
          cx(
            'ais-CurrentRefinements--noRefinement',
            classNames.noRefinementRoot
          ),
        props.className
      )}
    >
      <ul
        className={cx(
          'ais-CurrentRefinements-list',
          classNames.list,
          /* @MAJOR remove to ensure conformity with InstantSearch.css specs */
          !hasRefinements &&
            cx(
              'ais-CurrentRefinements-list--noRefinement',
              classNames.noRefinementList
            )
        )}
      >
        {items.map((item) => (
          <li
            key={[item.indexName, item.label].join('/')}
            className={cx('ais-CurrentRefinements-item', classNames.item)}
          >
            <span
              className={cx('ais-CurrentRefinements-label', classNames.label)}
            >
              {capitalize(item.label)}:{' '}
            </span>
            {item.refinements.map((refinement) => (
              <span
                key={refinement.label}
                className={cx(
                  'ais-CurrentRefinements-category',
                  classNames.category
                )}
              >
                <span
                  className={cx(
                    'ais-CurrentRefinements-categoryLabel',
                    classNames.categoryLabel
                  )}
                >
                  {refinement.label}
                </span>
                <button
                  type="button"
                  onClick={(event) => {
                    if (isModifierClick(event)) {
                      return;
                    }

                    item.refine(refinement);
                  }}
                  className={cx(
                    'ais-CurrentRefinements-delete',
                    classNames.delete
                  )}
                >
                  âœ•
                </button>
              </span>
            ))}
          </li>
        ))}
      </ul>
    </div>
  );
}
