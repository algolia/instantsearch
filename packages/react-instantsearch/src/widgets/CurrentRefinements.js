import connectCurrentRefinements from '../connectors/connectCurrentRefinements.js';
import CurrentRefinementsComponent from '../components/CurrentRefinements.js';

/**
 * The CurrentRefinements widget displays the list of filters currently applied to the search parameters.
 * It also lets the user remove each one of them.
 * @name CurrentRefinements
 * @kind widget
 * @propType {function} [transformItems] - If provided, this function can be used to modify the `items` provided prop of the wrapped component (ex: for filtering or sorting items). this function takes the `items` prop as a parameter and expects it back in return.
 * @themeKey ais-CurrentRefinements__root - the root div of the widget
 * @themeKey ais-CurrentRefinements__items - the container of the filters
 * @themeKey ais-CurrentRefinements__item - a single filter
 * @themeKey ais-CurrentRefinements__itemLabel - the label of a filter
 * @themeKey ais-CurrentRefinements__itemClear - the trigger to remove the filter
 * @themeKey ais-CurrentRefinements__noRefinement - present when there is no refinement
 * @translationKey clearFilter - the remove filter button label
 * @example
 * import React from 'react';
 *
 * import {ClearAll, RefinementList} from '../packages/react-instantsearch/dom';
 *
 * export default function App() {
 *   return (
 *     <InstantSearch
 *       className="container-fluid"
 *       appId="latency"
 *       apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *       indexName="ikea"
 *     >
 *       <CurrentRefinements />
 *       <RefinementList
          attributeName="colors"
          defaultRefinement={['Black']}
        />
 *     </InstantSearch>
 *   );
 * }
 */
export default connectCurrentRefinements(CurrentRefinementsComponent);
