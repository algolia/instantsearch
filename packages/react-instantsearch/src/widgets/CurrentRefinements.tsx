import React from 'react';
import { useCurrentRefinements } from 'react-instantsearch-core';

import { CurrentRefinements as CurrentRefinementsUiComponent } from '../ui/CurrentRefinements';

import type { CurrentRefinementsProps as CurrentRefinementsUiComponentProps } from '../ui/CurrentRefinements';
import type { UseCurrentRefinementsProps } from 'react-instantsearch-core';

type UiProps = Pick<
  CurrentRefinementsUiComponentProps,
  'items' | 'onRemove' | 'hasRefinements'
>;

export type CurrentRefinementsProps = Omit<
  CurrentRefinementsUiComponentProps,
  keyof UiProps
> &
  UseCurrentRefinementsProps;

export function CurrentRefinements({
  includedAttributes,
  excludedAttributes,
  transformItems,
  ...props
}: CurrentRefinementsProps) {
  const { items, refine, canRefine } = useCurrentRefinements(
    {
      includedAttributes,
      excludedAttributes,
      transformItems,
    },
    {
      $$widgetType: 'ais.currentRefinements',
    }
  );

  const uiProps: UiProps = {
    items,
    onRemove: refine,
    hasRefinements: canRefine,
  };

  return <CurrentRefinementsUiComponent {...props} {...uiProps} />;
}
