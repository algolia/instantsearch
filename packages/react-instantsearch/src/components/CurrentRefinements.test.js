/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';

import CurrentRefinements from './CurrentRefinements';

describe('CurrentRefinements', () => {
  it('renders a list of current refinements', () =>
    expect(
      renderer.create(
        <CurrentRefinements
          refine={() => null}
          items={[{
            label: 'Genre',
            value: 'clear all genres',
          }]}
          canRefine={true}
        />
      ).toJSON()
    ).toMatchSnapshot()
  );

  it('allows clearing unique items of a refinement', () =>
    expect(
      renderer.create(
        <CurrentRefinements
          refine={() => null}
          items={[{
            label: 'Genre',
            value: 'clear all genres',
            items: [{
              label: 'Sci-fi',
              value: 'clear sci-fi',
            }],
          }]}
          canRefine={true}
        />
      ).toJSON()
    ).toMatchSnapshot()
  );

  describe('Panel compatibility', () => {
    it('Should indicate when no more refinement', () => {
      const canRefine = jest.fn();
      const wrapper = mount(
        <CurrentRefinements
          refine={() => null}
          items={[{
            label: 'Genre',
            value: 'clear all genres',
            items: [{
              label: 'Sci-fi',
              value: 'clear sci-fi',
            }],
          }]}
          canRefine={true}
        />,
        {
          context: {canRefine},
          childContextTypes: {canRefine: React.PropTypes.func},
        },
      );

      expect(canRefine.mock.calls.length).toBe(1);
      expect(canRefine.mock.calls[0][0]).toEqual(true);
      expect(wrapper.find('.ais-CurrentRefinements__noRefinement').length).toBe(0);

      wrapper.setProps({canRefine: false});

      expect(canRefine.mock.calls.length).toBe(2);
      expect(canRefine.mock.calls[1][0]).toEqual(false);
      expect(wrapper.find('.ais-CurrentRefinements__noRefinement').length).toBe(1);
    });
  });
});
