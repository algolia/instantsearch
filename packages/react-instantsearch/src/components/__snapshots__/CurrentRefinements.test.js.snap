// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentRefinements allows clearing unique items of a refinement 1`] = `
<div
  className="ais-CurrentRefinements__root"
>
  <div
    className="ais-CurrentRefinements__items"
  >
    <div
      className="ais-CurrentRefinements__item ais-CurrentRefinements__itemParent"
    >
      <span
        className="ais-CurrentRefinements__itemLabel"
      >
        Genre
      </span>
      <div
        className="ais-CurrentRefinements__item"
      >
        <span
          className="ais-CurrentRefinements__itemLabel"
        >
          Sci-fi
        </span>
        <button
          className="ais-CurrentRefinements__itemClear"
          onClick={[Function]}
        >
          ✕
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`CurrentRefinements renders a list of current refinements 1`] = `
<div
  className="ais-CurrentRefinements__root"
>
  <div
    className="ais-CurrentRefinements__items"
  >
    <div
      className="ais-CurrentRefinements__item"
    >
      <span
        className="ais-CurrentRefinements__itemLabel"
      >
        Genre
      </span>
      <button
        className="ais-CurrentRefinements__itemClear"
        onClick={[Function]}
      >
        ✕
      </button>
    </div>
  </div>
</div>
`;
